### PR TITLE
Trainer Mode 

### PR DESCRIPTION
**Pull request by StarWitch**
For #193 .

Right now, the following is implemented:
* **Extra Lives (when needed)** - you can still die, but only lives that you lose when you go below 0 total lives will count as "extras". If you end up collecting extra lives, those get used first.
* **Extra Spellcards (when needed)** - same as above. Stored used first, extras given/tracked.
* **Invulnerability (with Hit Tracking)** - lets you turn on God Mode (like ``iddqd``) but will track whenever you get hit by a bullet.
* **No Powerdown When Hit** - if you die, you don't lose any of your Power gauge. 
* **HUD Stats & Tracking** - toggle whether or not you want to see stats in the HUD while playing the game. Some people can find them distracting. There is also a "per stage" and "per play-through" tracking extra lives/extra bombs/misses while invulnerable. That way you can see which stages trip you up the most, and how, and also overall stats for your entire run.

I broke it out into its own module so that additional functionality can be added easily enough down the line. This is a minimum viable implementation right now. 

The HUD display isn't the most polished but it can be changed easily. I put it in there just to see what it'd look like. I'm also hoping to eventually implement the ability to log when/where you end up dying, getting hit, or needing an extra bomb (which Spellcard, etc) and track stats on a menu entry somewhere.

I don't think adding a "Slowdown" feature is worth it at the moment, mostly because of how much work is being done on Coroutines and how screwy global timers are. We'd also have to slow down audio clips that play, which as far as I know isn't a thing that's implemented right now. And at the end of the day, either someone's using it to get better at the game (or STGs in general), or they're using it so they can see the whole game at their leisure, and in either case Slowdown doesn't really help them do that.

### CODE REVIEW

File: src/trainer.c [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    tnr->trainer_hits_stage = 0;
+
+    tnr->trainer_lives_total = 0;
+    tnr->trainer_bombs_total = 0;
+    tnr->trainer_hits_total = 0;
+}
+
+
+bool trainer_enabled(void) {
+    if (config_get_int(CONFIG_TRAINER_MODE) && config_get_int(CONFIG_TRAINER_STATS)) {
```
Comment (Akaricchi) [inline src/trainer.c]: I'm a bit confused about the `CONFIG_TRAINER_STATS` setting. What is it for exactly?

File: src/trainer.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+struct Trainer {
+	int trainer_lives_stage;
+	int trainer_bombs_stage;
+	int trainer_hits_stage;
+
+	int trainer_lives_total;
+	int trainer_bombs_total;
+	int trainer_hits_total;
+};
```
Comment (Akaricchi) [inline src/trainer.h]: I would recommend something like that instead:

```suggestion
struct Trainer {
	struct {
		int lives;
		int bombs;
		int hits;
	} total, stage;
};
```

Keeping total and stage stats in separate sub-structs like this allows easily resetting them separately with a memset call:
```c
memset(&trailer.stage, 0, sizeof(trainer.stage));
```

The `trainer_` prefix on all fields is also unnecessary.

File: src/global.h [hunk 3]
```diff
@@ -135,6 +136,8 @@ typedef struct {
 
 	StageInfo *stage;
 
+	Trainer tnr;
```
Comment (Akaricchi) [inline src/global.h]: The global object is bad enough; please rename this field to `trainer` for clarity.

File: src/trainer.c [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        if (config_get_int(CONFIG_TRAINER_LIVES) || config_get_int(CONFIG_TRAINER_BOMBS) || config_get_int(CONFIG_TRAINER_INVULN)) {
+            return 1;
```
Comment (Akaricchi) [inline src/trainer.c]: Please use the `true` and `false` macros for booleans.

File: src/player.c [hunk 5]
```diff
@@ -923,6 +938,12 @@ void player_death(Player *plr) {
 		return;
 	}
 
+	if (trainer_invulnerable_enabled()) {
+		trainer_append_hit_event(&global.tnr);
+		play_sound("shot3");
```
[the remark this answers is not in the document]
Comment (StarWitch) [reply src/player.c]: I was thinking of more of a "ping" sound. I'll try to find one that's more distinct but also low-key.

File: src/stagedraw.c [hunk 6]
```diff
@@ -1255,6 +1268,14 @@ static void stage_draw_hud_text(struct labels_s* labels) {
 	draw_label("Value:",       labels->y.value,   labels, &stagedraw.hud_text.color.label_value);
 	draw_label("Volts:",       labels->y.voltage, labels, &stagedraw.hud_text.color.label_voltage);
 	draw_label("Graze:",       labels->y.graze,   labels, &stagedraw.hud_text.color.label_graze);
+
+	if (trainer_enabled()) {
+		draw_label("Trainer Mode", labels->y.trainer, labels, &stagedraw.hud_text.color.label_trainer);
+		draw_label("Lives:", labels->y.trainer_lives, labels, &stagedraw.hud_text.color.label_trainer_lives);
+		draw_label("Spells:", labels->y.trainer_bombs, labels, &stagedraw.hud_text.color.label_trainer_bombs);
+		draw_label("Hits:", labels->y.trainer_hits, 0, &stagedraw.hud_text.color.label_trainer_hits);
```
[the remark this answers is not in the document]
Comment (StarWitch) [reply src/stagedraw.c]: Actually a typo here. Fixed anyways!

File: src/trainer.c (hunk 1 above)
Comment (StarWitch) [reply src/trainer.c]: It determines whether or not the stats are shown in real-time on the HUD, or kept strictly to the Stage Clear screen - I'm sure some people might find them distracting. However, its usage here wasn't quite correct, but I've fixed it.

File: src/player.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (StarWitch) [reply src/player.c]: OK, I added ``tick.opus``. It sounds distinct but isn't annoying, and is also CC-0/Public Domain. 

https://freesound.org/people/NenadSimic/sounds/268108/

File: src/trainer.c [hunk 7]
```diff
@@ -14,78 +14,92 @@
...
-
 bool trainer_enabled(void) {
-	if (config_get_int(CONFIG_TRAINER_MODE) && config_get_int(CONFIG_TRAINER_STATS)) {
-		if (config_get_int(CONFIG_TRAINER_LIVES) || config_get_int(CONFIG_TRAINER_BOMBS) || config_get_int(CONFIG_TRAINER_INVULN)) {
-			return 1;
-		}
+	if (config_get_int(CONFIG_TRAINER_MODE)) {
+		return true;
+	}
+	return false;
```
Comment (Akaricchi) [inline src/trainer.c]: Actually, why not just return the expression inside the `if` from those functions? Same with the new `player_can_bomb`, etc.

Another style nitpick that I missed: we don't put spaces before the opening parenthesis after `if`/`for`/`while`, sorry.

File: src/player.c [hunk 8]
```diff
@@ -850,19 +862,21 @@ void player_realdeath(Player *plr) {
...
 	if (!trainer_no_powerdown_enabled()) {
-		// if a player gets hit with trainer-no_pwrdn enabled...
+		// if "no powerdown" trainer is DISABLED, then powerdown as usual
 		int drop = max(2, (total_power * 0.15) / POWER_VALUE);
 		spawn_items(plr->deathpos, ITEM_POWER, drop);
 		player_set_power(plr, total_power * 0.7);
 		plr->voltage *= 0.9;
-		log_debug("Trainer Mode - No Powerdown");
+	} else {
+		// log when "no powerdown" trainer used
```
Comment (Akaricchi) [inline src/player.c]: The code is self-explanatory so I don't think this comment is needed.

**Comment by StarWitch** [comment]
Yeah, replays were going to be the next thing I worked on with respect to this. As I said, this isn't a "complete" implementation, it's a basic one to see how it looks and fits in with the game.

I think the idea of having to set which options you want before you start playing is a decent one, as it's also something I thought of, not only for replays but just with the spirit of the mode itself in general.

----

As for being able to see the good endings and all that... I think Trainer Mode should still allow it. But let me explain first.

As I said in the original Issue ticket, a lot of "hardcore" games are letting players do that these days. Celeste is my big inspiration for this in Taisei: it's an incredibly intense and competitive masocore platformer, where you're expected to die hundreds if not thousands of times in a casual playthrough, but you can still get to the end of the game and unlock everything with their Assist Mode turned on.

It encourages and incentivizes players to beat the game with Assist Mode turned off, though. It adds a badge to your completion screens saying that Assist Mode was used, which goes away when you beat each Chapter/Stage with it turned off. There's also messages throughout the game telling players that it's supposed to be challenging, and that the challenge is part of the genre (masocore). We could do the same thing with "Trainer Mode" enabled on the HUD as a mandatory feature (with only stats being toggleable), as well as having it marked as "Cheats Used" in replays, which is already a thing due to ``iddqd``.

Also, from a personal perspective, doing stage/spell practice is a different feel from playing through the game from start to finish. I play differently due to the flow of the stages and the feeling of momentum in a good run. It's hard to achieve that "zen" state when the start and stops are so sudden.

Anyways, that's just my opinion.

**Comment by Akaricchi** [comment]
Alright, I see where you're coming from. I do agree that it should cheat-mark the replays and display the HUD message. We can also track stage & game clears with trainer mode/cheats separately in the progress module, perhaps for a future stats screen. In that case I'd be ok with it.

**Comment by Akaricchi** [comment]
This system could be further extended later with options that not just make the game easier, but also harder, sillier, or just different somehow (various challenge modes, deadly pickups, permanent Mystia blindness, UFOs from th12, you name it). I actually thought of something like that before.

**Comment by StarWitch** [comment]
A challenge mode would be a lot of fun. A few of my friends actually complain that Touhou-themed STGs are "too easy" even on Lunatic so adding additional challenge options would be a cool idea just to show off.

**Comment by StarWitch** [comment]
I redid a bit of this to make it fit in better with the overall structure of the code. It also now works with replays, displays what replays had Trainer Mode enabled in the replay select menu, as well as displaying it in the character select. You also can't select it from in-game to prevent people flipping the settings during a run and potentially messing with replay stats - either you start a run with it, or you don't. It's not a particularly long game so I don't think this is too big of a deal.

**Comment by StarWitch** [comment]
I rebased this branch to get rid of the merge conflicts. Hopefully everything looks as it should now.

**Comment by StarWitch** [comment]
Closing this PR for now since we're going to have to think about rethinking the way game sessions are handled before v1.4, to make things easier to maintain.

**Pull request closed without merging**